src/utilities/masked_coeffs.jl

```diff
@@ -0,0 +1,29 @@
+"""
+    MaskedCoeffs(coefficients)
+
+A mask for coefficients. Supports `getindex(::MaskedCoeffs, ijk...)`
+that forwards to coefficients, and `getindex(::Type{MaskedCoeffs}, ijk...)`
+that forwards (at compile time) to the mask, which behaves as a
+BitArray of the coefficients.
+"""
+struct MaskedCoeffs{S, m, C}
+    coeffs::C
+    function MaskedCoeffs(coeffs::C) where {C}
+        m = BitArray(iszero.(coeffs))
+        return new{size(m), Tuple(m), C}(coeffs)
+    end
+end
```

Comment:
```suggestion
struct SparseCoeffs{S, C}
    coeffs::C
end
SparseCoeffs(coeffs) =
    SparseCoeffs{Tuple(findall(!iszero, coeffs)), typeof(coeffs)}(coeffs)
```
Storing the non-zero mask as a type parameter makes the implementation of `fused_increment` somewhat complicated, as requires searching through the mask for `true` entries to generate `fused_increment`. It would probably be better to store the indices of the non-zero coefficients as a type parameter. This would simplify the definition of `fused_increment` to something like
```
function fused_increment(u, dt, mc::SparseCoeffs{S}, tend) where {S}
    terms = Unrolled.unrolled_map(S) do nonzero_index
        i = nonzero_index isa CartesianIndex ? nonzero_index[2] : nonzero_index
        Base.Broadcast.broadcasted(*, dt * mc.coeffs[nonzero_index], tend[i])
    end
    Base.Broadcast.broadcasted(+, u, terms...)
end
```
I don't think we need to bother with generated functions for this.

Reply:
The main downside to what I wrote above is that it puts more data in type space. Since each index returned by `findall` is either an `Int64` or a `CartesianIndex` of 2 `Int64`s, the `sizeof` the `Tuple` in my example is either `4 * nnz` or `8 * nnz`, where `nnz` is the number of non-zero coefficients. On the other hand, each value in the `BitArray` is either `true` or `false`, so the `sizeof` the `Tuple` in the original version is `1 * N`, where `N` is the total number of coefficients.

This is easy to fix, though. We can just convert each index to either an `Int8` or a `Tuple{Int8, Int8}`, which will change the `sizeof` the `Tuple` in my example to either `1 * nnz` or `2 * nnz`. If fewer than half of the coefficients have non-zero values, this should put even less data into type space than the original version.

Reply:
I definitely like the name `SparseCoeffs` more, I'll update the name.

> Storing the mask as a type parameter makes the implementation of fused_increment somewhat complicated, as it requires searching through the mask for true entries to generate fused_increment.

It's not much of a lookup, it's as complicated as `CartesianIndex`, but I see your point, and I agree that your suggestion is more compact. Rather than storing all the `true`s/`false`s, we can store a tuple of `Int`s containing the indices with (e.g.) non-zero, and we can use `Int8` since the size of the coefficients will typically be small. I'll make the changes.

Reply:
I'll need to keep `S`, because `Tuple` flattens the result, and I don't want to rely on `findall` returning a particular element type to dispatch on between the 1D and 2D coefficients case.